### ISSUE TITLE
Fix false positives for array literal expressions

### ISSUE THREAD

**Issue by brandonkal** [stylelint/stylelint#4140]
> Clearly describe the bug

Stylelint is not giving special treatment for expressions inside styled components. This means that rules that use regex to locate css attribute brackets result in parseErrors.

> Which rule, if any, is the bug related to?

There are many. For instance: 'selector-attribute-brackets-space-inside': 'never'

> What CSS is needed to reproduce the bug?

```js
const Card = styled.div`
  &${[p.grid$]} {
   padding: 24px;
  }
`
```

> What stylelint configuration is needed to reproduce the bug?

e.g.

```json
{
  "rules": {
    "selector-attribute-brackets-space-inside": "never"
  }
}
```

> Which version of stylelint are you using?

10.1.0

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

CLI

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

css-in-js

> What did you expect to happen?

Stylelint should not attempt to parse array literals inside expressions. No parse errors should occur here. Arrays inside expressions are mistaken as css attribute brackets.

> What actually happened (e.g. what warnings or errors did you get)?

parseError

**Comment by brandonkal** [comment]
Here is a list of rules with false positives because Stylelint is attempting to read inside template expressions:

- [ ] selector-attribute-brackets-space-inside
- [ ] selector-attribute-operator-space-before
- [ ] selector-attribute-operator-space-after
- [ ] function-whitespace-after
- [ ] declaration-bang-space-before
- [ ] function-name-case
- [ ] no-descending-specificity

**Comment by brandonkal** [comment]
I have found some more broken rules:

- [ ] selector-combinator-space-after
- [ ] selector-combinator-space-before
- [ ] selector-descendant-combinator-no-non-space
- [ ] selector-pseudo-class-parentheses-space-inside

Here you can see my config. Note the long list of rules that had to be disabled due to this issue that remains:
```js
module.exports = {
  extends: ['stylelint-config-standard', 'stylelint-config-prettier'],
  plugins: ['@brandonkal/stylelint-csstree-validator'],
  rules: {
    'comment-empty-line-before': null,
    'declaration-empty-line-before': null,
    'csstree/validator': true,
    'unit-no-unknown': [
      true,
      {
        ignoreUnits: ['emin', 'emax', 'eh', 'ew'],
      },
    ],
    // Fixes for errors in Stylelint rules
    'selector-attribute-brackets-space-inside': null,
    'selector-attribute-operator-space-before': null,
    'selector-attribute-operator-space-after': null,
    'function-whitespace-after': null,
    'declaration-bang-space-before': null,
    'function-name-case': null,
    'no-descending-specificity': null,
    'selector-combinator-space-after': null,
    'selector-combinator-space-before': null,
    'selector-descendant-combinator-no-non-space': null,
    'selector-pseudo-class-parentheses-space-inside': null,
    // From linaria
    'property-no-vendor-prefix': true,
    'string-no-newline': true,
    'value-no-vendor-prefix': true,
    'no-empty-source': null,
    'no-extra-semicolons': null,
    'no-missing-end-of-source-newline': null,
    'selector-pseudo-class-no-unknown': [
      true,
      {
        ignorePseudoClasses: ['global'],
      },
    ],
  },
}
```

Could someone please label this as a bug?

**Comment by evilebottnawi** [comment]
Can you provide example of code with bugs ruls?

**Comment by brandonkal** [comment]
See original post. Essentially any interpolation with an array literal inside.

**Comment by evilebottnawi** [comment]
Thanks for issue, PR welcome :+1:

**Comment by jeddy3** [comment]
> Thanks for issue, PR welcome

For anyone picking this up, you can see if this can be fixed in the [postcss-jsx parser](https://www.npmjs.com/package/postcss-jsx) or whether it's best to add array literal expressions heuristics to the`isStandardSyntax*` utils (e.g. [isStandardSyntaxSelector](https://github.com/stylelint/stylelint/blob/master/lib/utils/isStandardSyntaxSelector.js)) here in stylelint.

**Comment by jeddy3** [comment]
We've forked the parser we use for CSS-in-JS [into stylelint organisation](https://github.com/stylelint/postcss-css-in-js), and we're looking for help to fix our fork to close this issue.

See https://github.com/stylelint/stylelint/issues/4574.

**Comment by jeddy3** [comment]
Closing as the [`syntax` option was removed](https://github.com/stylelint/stylelint/issues/5289) for the [14.0.0 release](https://github.com/stylelint/stylelint/issues/5205). (See the [migration guide](https://github.com/stylelint/stylelint/blob/main/docs/migration-guide/to-14.md#syntax-option-and-automatic-inferral-of-syntax).)

We'll also [be deprecating](https://github.com/stylelint/postcss-css-in-js/issues/225) the [forked CSS-in-JS package](https://github.com/stylelint/postcss-css-in-js), in favour of leaner custom syntaxes.

Please consider [writing a custom syntax](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/syntaxes.md) specific to the CSS-in-JS library in this issue and fixing the bug there.
